### DISCUSSION TITLE
Error importing TypeScript file into the configuration file - unknown file extension

### DISCUSSION THREAD

**Discussion by Ixluxi** [cypress-io/cypress#30228] [Questions and Help]
Hi,

I have a package installed (one I created) that contains a TypeScript file called `test-data-utilities.ts`

In my config file (`qa.config.ts`), I try to import it, but it imports (auto-completes) as `.js`...

`import TestDataUtilities from '@mypackage/cypress-test-data-utilities/test-data-utilities.js'`

So I switch '.js' to '.ts' and try to use the class:

`const testDataUtilities = new TestDataUtilities()`

I get error:

```
Your configFile is invalid: C:\Users\me\Desktop\automation\projects\ProjectName\FrameworkName\cypress\configurations\qa.config.ts

It threw an error when required, check the stack trace below:

TypeError: Unknown file extension ".ts" for C:\Users\me\Desktop\automation\projects\ProjectName\FrameworkName\cypress\configurations\qa.config.ts
```

I have TypeScript installed in the framework, but does the package itself need to have TypeScript installed? (typescript is not present in the package's list of dependencies).

Thank you for any help you can provide.

EDIT: The work-around is to change my package file to js or bake the methods into each config file, which defeats the point of using TypeScript.

**Comment by portyu9** [comment]
Hi @Ixluxi 

For a reusable package, I would **not** make consumers import the package's raw `.ts` source.

The supported packaging pattern is to compile the package to JavaScript and publish the generated `.d.ts` files for TypeScript consumers.

For example:

```json
{
  "scripts": {
    "build": "tsc -p tsconfig.json"
  },
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ]
}
```

and:

```json
{
  "compilerOptions": {
    "declaration": true,
    "outDir": "dist"
  }
}
```

Then consumers import from the package normally, without replacing `.js` with `.ts`.

Adding `typescript` as a **runtime dependency** of the package is not what makes Node able to execute raw TypeScript. `typescript` normally belongs in the package's development/build dependencies.

There is also an important version difference from when this question was posted: current Cypress supports TypeScript config files and, since Cypress 15, uses `tsx` to run them. That improves Cypress's own config-file handling, but it does not make exposing raw `.ts` from a reusable package the recommended runtime contract.

If this is a monorepo where you intentionally consume source TypeScript directly, that can be configured through your workspace/build tooling, but for a package intended to be reused or published, build it to JavaScript first.

No screenshots or attachments were included in this discussion.

Please mark this answer as accepted if it helped, thank you.
